File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatadetail/WeatherDataDetailPresenterImpl.java
package pl.mosenko.sunnypodlaskie.mvp.weatherdatadetail;

import android.os.Bundle;
import android.support.annotation.NonNull;

import com.hannesdorfmann.mosby3.mvp.MvpBasePresenter;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import pl.mosenko.sunnypodlaskie.ApplicationPodlaskieWeather;
import pl.mosenko.sunnypodlaskie.persistence.dao.WeatherDataEntityDAO;
import pl.mosenko.sunnypodlaskie.persistence.entities.WeatherDataEntity;
import pl.mosenko.sunnypodlaskie.util.WeatherDataDetailPresentationModelConverter;

import static pl.mosenko.sunnypodlaskie.mvp.weatherdatadetail.WeatherDataDetailFragment.ARG_WEATHER_DATA_ID;

/**
 * Created by syk on 06.06.17.
 */

public class WeatherDataDetailPresenterImpl extends MvpBasePresenter<WeatherDataDetailContract.View> implements  WeatherDataDetailContract.Presenter{

    @Inject
    WeatherDataEntityDAO weatherDataEntityDAO;
    private CompositeDisposable compositeDisposable;

    public WeatherDataDetailPresenterImpl() {
        injectFields();
        initializeCompositeDisposable();
    }

    private void injectFields() {
        ApplicationPodlaskieWeather.sharedApplication().getDIComponent().inject(this);
    }

    private void initializeCompositeDisposable() {
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void onViewCreated(Bundle savedInstanceState) {
            if (savedInstanceState.containsKey(ARG_WEATHER_DATA_ID)) {
                Long weatherDataId = savedInstanceState.getLong(ARG_WEATHER_DATA_ID);
                queryForWeatherDataById(weatherDataId);
            }
    }

    private void queryForWeatherDataById(@NonNull Long weatherDataId) {
        if (!isViewNotNullAttached()) {
         return;
        }
        safelyDisposeRepositorySubscription();
        Disposable disposable = weatherDataEntityDAO.rxQueryForId(weatherDataId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map(this::formatWeatherDataEntityToDisplay)
                .subscribe(getView()::loadData, getView()::showError);
        compositeDisposable.add(disposable);
    }

    private boolean isViewNotNullAttached() {
        return getView() != null && isViewAttached();
    }

    private WeatherDataDetailPresentationModel formatWeatherDataEntityToDisplay(WeatherDataEntity weatherDataEntity) {
        return WeatherDataDetailPresentationModelConverter.convert(weatherDataEntity);
    }

    @Override
    public void detachView(boolean retainInstance) {
        super.detachView(retainInstance);
        safelyDisposeRepositorySubscription();
    }

    private void safelyDisposeRepositorySubscription() {
        if (compositeDisposable != null && compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
    }
}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/mvp/weatherdatadetail/WeatherDataDetailPresenterImpl.kt
package pl.mosenko.sunnypodlaskie.mvp.weatherdatadetail

import android.os.Bundle
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.runBlocking
import pl.mosenko.sunnypodlaskie.mvp.MvpBasePresenter
import pl.mosenko.sunnypodlaskie.persistence.dao.WeatherDataEntityDao
import pl.mosenko.sunnypodlaskie.persistence.entities.WeatherDataEntity
import pl.mosenko.sunnypodlaskie.util.WeatherDataDetailPresentationModelConverter

/**
 * Created by syk on 06.06.17.
 */
class WeatherDataDetailPresenterImpl(
    var weatherDataEntityDao: WeatherDataEntityDao,
) : MvpBasePresenter<WeatherDataDetailContract.View>(), WeatherDataDetailContract.Presenter {

    private var compositeDisposable: CompositeDisposable? = null

    init {
        initializeCompositeDisposable()
    }

    private fun initializeCompositeDisposable() {
        compositeDisposable = CompositeDisposable()
    }

    override fun onViewCreated(savedInstanceState: Bundle) {
        if (savedInstanceState.containsKey(WeatherDataDetailFragment.ARG_WEATHER_DATA_ID)) {
            val weatherDataId =
                savedInstanceState.getLong(WeatherDataDetailFragment.ARG_WEATHER_DATA_ID)
            queryForWeatherDataById(weatherDataId)
        }
    }

    private fun queryForWeatherDataById(weatherDataId: Long) {
        if (!isViewNotNullAttached()) {
            return
        }
        safelyDisposeRepositorySubscription()
        val disposable =
            Observable.fromCallable {
                runBlocking {
                    weatherDataEntityDao.getWeatherDataById(weatherDataId)
                }
            }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .map { weatherDataEntity: WeatherDataEntity ->
                    formatWeatherDataEntityToDisplay(
                        weatherDataEntity
                    )
                }
                .subscribe({ weatherDataDetailPresentationModel: WeatherDataDetailPresentationModel ->
                    view!!.loadData(
                        weatherDataDetailPresentationModel
                    )
                }) { throwable: Throwable -> view!!.showError(throwable) }
        compositeDisposable?.add(disposable)
    }

    private fun isViewNotNullAttached(): Boolean {
        return view != null && isViewAttached
    }

    private fun formatWeatherDataEntityToDisplay(weatherDataEntity: WeatherDataEntity): WeatherDataDetailPresentationModel {
        return WeatherDataDetailPresentationModelConverter.convert(weatherDataEntity)
    }

    override fun detachView() {
        super.detachView()
        safelyDisposeRepositorySubscription()
    }

    private fun safelyDisposeRepositorySubscription() {
        if (compositeDisposable?.isDisposed == true) {
            compositeDisposable!!.dispose()
        }
    }
}
